### PR TITLE
[Sprint 52] 1.2.x only XD-3125, XD-3126 Refactor Codec configuration.

### PR DESCRIPTION
**Pull request by dturanski**
This is a redo of https://github.com/spring-projects/spring-xd/pull/1717 addressing @markpollack review comments.

### CODE REVIEW

File: spring-xd-codec/src/main/java/org/springframework/xd/dirt/integration/bus/serializer/kryo/AbstractKryoCodec.java [hunk 1]
```diff
@@ -38,17 +38,6 @@
  */
 public abstract class AbstractKryoCodec<T> extends AbstractCodec<T> {
 
-	protected final static int FILE_REGISTRATION_ID = 40;
```
Comment (markpollack) [inline spring-xd-codec/src/main/java/org/springframework/xd/dirt/integration/bus/serializer/kryo/AbstractKryoCodec.java]: The intention here is that these would be done by the user should they want to have IDs instead of strings for classnames for these types?

File: spring-xd-codec/src/main/java/org/springframework/xd/dirt/integration/bus/serializer/kryo/AbstractKryoCodec.java [hunk 2]
```diff
@@ -38,17 +38,6 @@
  */
 public abstract class AbstractKryoCodec<T> extends AbstractCodec<T> {
```
Comment (markpollack) [inline spring-xd-codec/src/main/java/org/springframework/xd/dirt/integration/bus/serializer/kryo/AbstractKryoCodec.java]: Related to the optimization work for serialization/deserialization that would go into another PR, AbstractCode always using a ByteArrayInputStream might not be a good idea.  Perhaps we should change this abstract base class to an interface.

Comment (markpollack) [inline spring-xd-codec/src/main/java/org/springframework/xd/dirt/integration/bus/serializer/kryo/AbstractKryoCodec.java]: AbstractKryoCodec overrides deserialize(byte[] bytes) with the same exact implementation as the base class, will remove on merge.

**Comment by markpollack** [comment]
Do we want this also for the 1.2.x branch?

**Comment by dturanski** [comment]
Only for 1.2 branch

**Comment by markpollack** [comment]
but also on master?

Take a look at https://github.com/markpollack/spring-xd/tree/dturanski-XD-3125b and the rework in AbstractKryoMultiTypeCodec to see if you are ok with it.

**Comment by dturanski** [comment]
I thought master was for 2.0. All this will move to spring-bus

**Comment by markpollack** [comment]
merged on 1.2.x as bc15d3861732a618d61c8e96508469e9afcd47ae
